File: hackerrank/problem solving/algorithms/2 - implementation/4 - between two sets/between_two_sets.ts
// https://www.hackerrank.com/challenges/between-two-sets/problem?isFullScreen=true

function main() {
  readLine();
  const a: number[] = readNumbers();
  const b: number[] = readNumbers();
  console.log(betweenTwoSets(a, b));
}

function readNumbers(): number[] {
  return readLine()
    .split(' ')
    .map((x) => +x);
}

// n1, n2: length of a, b
// k: number of multiples tested = gcdOfB / lcmOfA
// T: O(n1 + n2 + k)
// S: O(1) extra space
function betweenTwoSets(a: number[], b: number[]): number {
  const lcmOfA: number = a.reduce(lcm);
  const gcdOfB: number = b.reduce(gcd);

  let count = 0;
  for (let i = lcmOfA; i <= gcdOfB; i += lcmOfA) {
    if (gcdOfB % i === 0) count++;
  }
  return count;
}

function lcm(a: number, b: number): number {
  return (a * b) / gcd(a, b);
}

function gcd(a: number, b: number): number {
  while (b !== 0) {
    const t = b;
    b = a % b;
    a = t;
  }
  return a;
}

//////////////////////////////////////////////////

import readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
let inputLines: string[] = [];
rl.on('line', (line: string) => inputLines.push(line));
rl.on('close', main);

function readLine(): string {
  return inputLines.shift()!;
}

File: hackerrank/problem solving/algorithms/2 - implementation/4 - between two sets/between_two_sets.rb
# https://www.hackerrank.com/challenges/between-two-sets/problem?isFullScreen=true

def main
  gets
  a = read_numbers
  b = read_numbers
  print(between_two_sets(a, b))
end

def read_numbers
  gets.split.map!(&:to_i)
end

# n1, n2: length of a, b
# k: number of multiples tested = gcd_of_b / lcm_of_a
# T: O(n1 + n2 + k)
# S: O(1) extra space
def between_two_sets(a, b)
  lcm_of_a = a.reduce(:lcm)
  gcd_of_b = b.reduce(:gcd)
  (lcm_of_a..gcd_of_b).step(lcm_of_a).count { (gcd_of_b % _1).zero? }
end

main if __FILE__ == $0
